### DISCUSSION TITLE
Best practice for log in as user/impersonate user?

### DISCUSSION THREAD

**Discussion by louisbarclay** [supabase/supabase#2331] [Questions] [answered]
Hi,

I couldn't find anything about this in the documentation.

I have a user whose dashboard is broken. I'd like to log in as them so I can debug.

What's the best way of achieving this?

**Comment by burggraf** [comment]
There are a couple ways to achieve this:

1. If the user can change their password to something temporary and share that password with you, you can log in with that temporary password.  This is a low tech solution and not very secure, but should work.
2. You could create a custom JWT token for the user and use the following functions to create a session based on that user's info: [setSession()](https://github.com/supabase/gotrue-js/blob/4575b2a116048b8226b5d4bd12a33777571b9b2a/src/GoTrueClient.ts#L271) [setAuth()](https://github.com/supabase/gotrue-js/blob/4575b2a116048b8226b5d4bd12a33777571b9b2a/src/GoTrueClient.ts#L301)
3. There's a new feature coming out in about a week (maybe less) that'll let you intercept a generated magic link -- you could then redirect that magic link to your own email, and you could use that link to log on as that user.  This sounds like the easiest, most flexible way to do it.

**Reply by louisbarclay** [reply]
Awesome, thanks so much! Option 3 sounds great, I will wait for that before building option 2 which I was considering.

**Reply by louisbarclay** [reply]
Did this new feature get released? I wasn't sure where to go to find the release notes to figure this out for myself

**Reply by burggraf** [reply]
Sorry, they did not release it yet, but it's definitely going to be released next week, along with a lot of great new stuff!

**Reply by burggraf** [reply]
Watch: https://github.com/supabase/gotrue/pull/133

**Comment by awalias** [comment, marked as answer]
we now have this in supabase-js , you will need to start the client with your `service_role` key and use `supabase.auth.api.generateLink`:

https://supabase.io/blog/2021/07/28/supabase-auth-passwordless-sms-login#generate-confirmation-links

**Reply by louisbarclay** [reply]
This did not work for me.

I tried this code:

```
export const getMagicLink = async (email) => {
  return await supabaseAdmin.auth.api.generateLink({ email, type: "invite" });
};
```

Received this response:

`{"data":null,"error":{"message":"Could not read body: json: cannot unmarshal object into Go struct field GenerateLinkParams.type of type string","status":400}}`
